### PR TITLE
Optimize structs

### PR DESCRIPTION
**Pull request by jcchavezs**
This PR detect structs that would use less memory if their fields were sorted.

```shell
$betteralign -apply ./...   
                               
~/Workspace/source/github.com/corazawaf/coraza/debuglog/default.go:14:19: struct with 24 pointer bytes could be 16
~/Workspace/source/github.com/corazawaf/coraza/debuglog/default.go:90:20: struct with 32 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/types/waf.go:142:19: struct with 40 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:24:18: struct with 128 pointer bytes could be 96
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:46:26: struct with 40 pointer bytes could be 32
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:78:30: struct with 32 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:94:18: struct with 152 pointer bytes could be 104
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:112:16: struct with 104 pointer bytes could be 88
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:129:27: struct with 64 pointer bytes could be 56
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:139:23: struct with 24 pointer bytes could be 16
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:146:24: struct with 32 pointer bytes could be 16
~/Workspace/source/github.com/corazawaf/coraza/auditlog/auditlog.go:152:20: struct with 200 pointer bytes could be 104
~/Workspace/source/github.com/corazawaf/coraza/auditlog/concurrent_writer.go:19:23: struct with 64 pointer bytes could be 48
~/Workspace/source/github.com/corazawaf/coraza/auditlog/logger.go:13:13: struct of size 56 could be 48
~/Workspace/source/github.com/corazawaf/coraza/auditlog/serial_writer.go:16:19: struct with 104 pointer bytes could be 80
~/Workspace/source/github.com/corazawaf/coraza/rules/operator.go:9:22: struct with 64 pointer bytes could be 48
~/Workspace/source/github.com/corazawaf/coraza/internal/corazarules/rule.go:12:19: struct with 160 pointer bytes could be 104
~/Workspace/source/github.com/corazawaf/coraza/internal/corazarules/rule_match.go:17:16: struct with 64 pointer bytes could be 56
~/Workspace/source/github.com/corazawaf/coraza/internal/corazarules/rule_match.go:52:18: struct with 144 pointer bytes could be 120
~/Workspace/source/github.com/corazawaf/coraza/internal/collections/named.go:88:27: struct with 16 pointer bytes could be 8
~/Workspace/source/github.com/corazawaf/coraza/macro/macro.go:37:17: struct with 32 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/internal/corazawaf/body_buffer.go:19:17: struct with 48 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/internal/corazawaf/body_buffer.go:96:23: struct with 16 pointer bytes could be 8
~/Workspace/source/github.com/corazawaf/coraza/internal/corazawaf/rule.go:24:23: struct with 32 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/internal/corazawaf/rule.go:45:28: struct with 24 pointer bytes could be 16
~/Workspace/source/github.com/corazawaf/coraza/internal/corazawaf/rule.go:58:25: struct with 40 pointer bytes could be 32
~/Workspace/source/github.com/corazawaf/coraza/internal/corazawaf/rule.go:94:11: struct of size 328 could be 320
~/Workspace/source/github.com/corazawaf/coraza/internal/corazawaf/transaction.go:40:18: struct of size 944 could be 912
~/Workspace/source/github.com/corazawaf/coraza/internal/corazawaf/waf.go:34:10: struct of size 472 could be 448
~/Workspace/source/github.com/corazawaf/coraza/internal/transformations/normalise_path_win.go:146:14: struct with 56 pointer bytes could be 40
~/Workspace/source/github.com/corazawaf/coraza/internal/actions/ctl.go:46:12: struct with 40 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/internal/actions/expirevar.go:15:18: struct with 32 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/internal/actions/initcol.go:13:16: struct with 32 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/internal/actions/setenv.go:15:15: struct with 32 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/internal/operators/rbl.go:20:10: struct with 24 pointer bytes could be 16
~/Workspace/source/github.com/corazawaf/coraza/internal/seclang/directives.go:26:23: struct with 216 pointer bytes could be 208
~/Workspace/source/github.com/corazawaf/coraza/internal/seclang/parser.go:24:13: struct with 64 pointer bytes could be 48
~/Workspace/source/github.com/corazawaf/coraza/internal/seclang/parser.go:214:19: struct with 144 pointer bytes could be 120
~/Workspace/source/github.com/corazawaf/coraza/internal/seclang/rule_parser.go:22:17: struct with 56 pointer bytes could be 40
~/Workspace/source/github.com/corazawaf/coraza/internal/seclang/rule_parser.go:305:18: struct with 208 pointer bytes could be 200
~/Workspace/source/github.com/corazawaf/coraza/config.go:102:16: struct of size 136 could be 128
~/Workspace/source/github.com/corazawaf/coraza/config.go:208:21: struct with 48 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/http/interceptor.go:21:20: struct with 48 pointer bytes could be 40
~/Workspace/source/github.com/corazawaf/coraza/testing/profile/profile.go:7:11: struct with 48 pointer bytes could be 40
~/Workspace/source/github.com/corazawaf/coraza/testing/profile/profile.go:15:17: struct with 128 pointer bytes could be 112
~/Workspace/source/github.com/corazawaf/coraza/testing/profile/profile.go:36:15: struct with 280 pointer bytes could be 264
~/Workspace/source/github.com/corazawaf/coraza/testing/profile/profile.go:51:14: struct with 88 pointer bytes could be 80
~/Workspace/source/github.com/corazawaf/coraza/testing/profile/profile.go:59:21: struct with 136 pointer bytes could be 112
~/Workspace/source/github.com/corazawaf/coraza/testing/profile/profile.go:72:27: struct with 40 pointer bytes could be 24
~/Workspace/source/github.com/corazawaf/coraza/testing/engine.go:19:11: struct with 344 pointer bytes could be 304
```

PENDING: benchmarks comparison

cc @anuraaga 

Question: shall we include this tool in the format task?

### CODE REVIEW

**Review by M4tteoP — APPROVED** [review]
Nice! 

> Question: shall we include this tool in the format task?

It seems reasonable to me

File: auditlog/logger.go [hunk 1]
```diff
@@ -11,20 +11,20 @@ import (
...
 	File string
 
-	// FileMode is the mode to use when creating File.
-	FileMode fs.FileMode
-
 	// Dir is the path to the directory to write formatted audit logs to.
 	Dir string
 
+	// FileMode is the mode to use when creating File.
+	FileMode fs.FileMode
```
Comment (anuraaga) [inline auditlog/logger.go:24]: Separating mode from what it applies to is an example of maintainability / readability problems introduced by reordering

File: internal/corazarules/rule_match.go [hunk 2]
```diff
@@ -25,6 +23,8 @@ type MatchData struct {
 	Message_ string
 	// Macro expanded logdata
 	Data_ string
+	// Variable
+	Variable_ variables.RuleVariable
```
Comment (anuraaga) [inline internal/corazarules/rule_match.go:27]: MatchData is one where it is potentially high impact as we could allocate a lot of them per request. However, separating Variable_ from Key_/Value_ seems like a large loss in readability. I would lean towards favoring the readability

File: internal/corazarules/rule_match.go [hunk 3]
```diff
@@ -50,6 +50,7 @@ func (m *MatchData) Data() string {
 // MatchedRule contains a list of macro expanded messages,
 // matched variables and a pointer to the rule
 type MatchedRule struct {
```
Comment (anuraaga) [inline internal/corazarules/rule_match.go:52]: This change seems OK readability wise

**Review by anuraaga — COMMENTED** [review]
I don't think this should be automated for sure. Code maintainability can go way down because of the ordering of fields. Of course, without automation, in the long term we won't have optimized memory layout anymore as things change.

Using this as a hint to find high-impact changes could be good, but this needs a lot of care. I have added some comments on examples of significant losses in readability. Going through, I could really only find a couple of examples where the reordering seems reasonable code-wise and is actually on the hot path (singletons don't need packing). Efficiency is important, but reliability is more so and that's where code readability is actually an issue (just a guess, but I wouldn't be surprised of `OutboundErrorData` and `OutboundDataError` issue came up due to a readability problem).

For context:

Go can't change default behavior, an opt-in may happen some day (this would be best)
https://github.com/golang/go/issues/10014

And the lack of associated PRs makes me suspect even TiDB isn't making such changes
https://github.com/pingcap/tidb/issues/9674

**Comment by jcchavezs** [comment]
Thanks for the insightful comments @anuraaga. I think the readability
concern is indeed important. I think I will close this PR as even if we
apply the reordering where it makes sense it will be really easy to forget
and change it again somehow so not worth.

**Comment by jptosso** [comment]
I'm with @anuraaga. Most structs variables are sorted in a way that enhances readability. If it doesn't provide an important performance improvement, I wouldn't implement it. Maybe we could add a "blacklist" of structs not to modify

**Pull request closed without merging**